lib/context.ts

```diff
@@ -30,7 +30,17 @@ export function toContext(vocab: Vocab): string {
     // Generation of a unit for properties
     const propertyContext = (property: RDFProperty, for_class = true): Context|string => {
         // the real id of the property...
-        const url = `${global.vocab_url}${property.id}`
+        let url = `${global.vocab_url}${property.id}`;
+        // if defined_by is set, we have a foreign property definition, so use
+        // its URL instead.
+        if ("defined_by" in property) {
+            // possibly expand prefixes
+            for (let prefix of vocab.prefixes) {
+                if (property.defined_by.indexOf(`${prefix.prefix}:`) > -1) {
+                    url = property.defined_by.replace(`${prefix.prefix}:`, prefix.url);
+                }
+            }
+        }
```

Comment:
I think I understand what you intend to do here, although I would like to see a 'before' and 'after' example to be sure.

However... I do not see the utility, _for our usage in VC,_ of this change. The role of the `defined_by` URL in both the VCDM and the DI vocabularies is to _link back to the specification where the term is defined._ It was a conscious choice to say that the vocabulary file does not include any normative definition of a term, because that must be done in the corresponding recommendation.

If our goal is to evolve the tool toward a more useful tool for vocabulary management out there, it becomes a totally different project. It is a worthy one, for sure, but I would prefer to fork the repo and doing it from there; mixing in such developments into the VC work may result in a lot of extra changes, PRs, etc., and I do not think we should be doing that right now.

But I may misunderstand your goals.


---

The other changes for the js versions are irrelevant at this point. They are "just" the generated files...

Reply:
Currently, it's not possible to use this tool to generate a context file that contains "foreign properties" (that may be the wrong phrase for this...). I'm currently trying to target expressing this context file via `yml2vocab`, but couldn't because it makes heavy use of schema.org terms:
https://github.com/digitalbazaar/citizenship-vocab/blob/main/contexts/citizenship-v2.jsonld

That's the main thing I'm after with this PR, so if there's a better / correct way to do it, I'm all ears. 😁

Reply:
Hm. I think I understand where you want to go, but I do not think your changes would do it.

The problem is, as I see, is that you set up goals (worthy as they are!) that goes a little bit beyond what this was meant to do. The yml file is currently aimed at a vocabulary at a single namespace URL. The terms will automatically get the URL based on the "global" setting of the vocabulary, and everything is centered around this. You can _refer_ to terms in other vocabularies for the domain/range/subClass setting, which all rely on a namespace prefix. But you cannot make a _statement_ on a term that is not in this vocabulary (e.g., you cannot make a statement on the `name` property in the VCDM vocabulary, because that is in schema.org, or on the `proof`  property, because that is defined in a separate vocabulary as well). That also means that the extra tool in the package that generates a context file will of course be partial: it creates the context terms only for, say, the VCDM terms.

That is why I was saying above that:

> If our goal is to evolve the tool toward a more useful tool for vocabulary management out there, it becomes a totally different project. It is a worthy one, for sure, but I would prefer to fork the repo and doing it from there; mixing in such developments into the VC work may result in a lot of extra changes, PRs, etc., and I do not think we should be doing that right now.

I am just worried about creating issues for the WG's work, that is all.

---
B.t.w., if your goal is to create a really useful `@context` file, there are other things that the vocabulary cannot express. For example, RDFS/OWL cannot formally express that the value of a property is a separate graph; this had to be added to the vocabulary in English prose. On the other hand, that is something the context can do using the container feature. That should also be "added" to the project...